l3build.lua

```diff
@@ -24,50 +24,207 @@ for those people who are interested.
-build_kpse_path = match(lookup("l3build.lua"),"(.*[/])")
-local function build_require(s)
-  require(lookup("l3build-"..s..".lua", { path = build_kpse_path } ) )
+
+-- # Start of the booting process
+
+local is_main  -- Whether the script is called first
+local work_dir -- the directory containing "build.lua" and friends
+
+assert(not _G.l3build, "No self call")
+
+---@alias flag_table_t table<string, boolean>
+
+---@class l3build_t
+---@field debug flag_table_t the special --debug-foo CLI arguments
+---@field PACKAGE string "l3build", `package.loaded` key
+---@field NAME string "l3build", display name
+---@field PATH string synonym of `launch_dir` .. "/l3build.lua"
+---@field work_dir string where the "build.lua" lives
+---@field launch_dir string where "l3build.lua" and friends live
+---@field start_dir string the current directory at load time
+local l3build = { -- global data available as package.
+  debug = {} -- storage for special debug flags (private UI)
+}
+
+do
+  -- the directory containing "l3build.lua" by kpse
+  local kpse_dir = match(kpse.lookup("l3build.lua"), ".*/")
+
+  -- Setup dirs where require will look for modules.
+
+  local launch_dir -- the directory containing "l3build.lua"
+
+  -- File operations are aided by the LuaFileSystem module
+  local lfs = require("lfs")
+
+  local start_dir = lfs.currentdir() .. "/" -- this is the current dir at launch time
+
+  ---Extract dir and base from path
+  ---@param path string
+  ---@return string dir includes a trailing '/', defaults to "./"
+  ---@return string base
+  local function to_dir_base(path)
+    local dir, base = match(path, "(.*/)(.*)")
+    if not dir then dir, base = "./", path end
+    return dir, base
+  end
+
+  ---Central function to allow launching l3build from a subdirectory
+  ---of a local repository.
+  ---Find `base` in `dir` or one of its parents, returns the container.
+  ---We do not assume that `dir` contains no ".." component
+  ---such that we cannot use `to_dir_base`.
+  ---Instead we append `/..` and rely on lua to do the job.
+  ---The max number of trials is the number of components
+  ---of the absolute path of `dir`, which majorated in the for loop
+  ---Intermediate directories must exist.
+  ---@param dir string must end with '/'
+  ---@param base string relative file or directory name
+  ---@return string|nil dir ends with '/' when non nil
+  local function container(dir, base)
+    for _ in gmatch(dir .. lfs.currentdir(), "[^/]+") do
+      local p = dir .. base
+      if os.rename(p, p) then return dir end -- true iff file or dir at the given path
+      -- synonym of previous line:
+      -- if package.searchpath("?", p, "", "") then return dir end
+      dir = dir .. "../"
+    end
+  end
+
+  local cmd_path = arg[0]
+  local cmd_dir, cmd_base = to_dir_base(cmd_path)
+
+  is_main = cmd_base == "l3build" or cmd_base == "l3build.lua"
+  assert(is_main == not not (match(arg[0], "l3build$") or match(arg[0], "l3build%.lua$")))
+  -- l3b_dir:
```

Comment:
The previous branching command is at line #117.
Do "foo-l3build" and "foo-l3build.lua" make any sense ?